File: Java/Daily/Medium/MaxNumberOfDistinctAfterOperations.java
import java.util.*;
// 3397
class Solution {
    public int maxDistinctElements(int[] nums, int k) {

        int last_picked = Integer.MIN_VALUE;
        Arrays.sort(nums);
        int distinct = 0;
        for (int num: nums) {
            int low = num - k;
            int high = num + k;

            if (last_picked < low)
                last_picked = low;
            else
                last_picked += 1;
            
            if (last_picked <= high)
                distinct++;
            else
                last_picked -= 1;
        }

        return distinct;
    }
}

File: Daily/MaxNumberOfDistinctAfterOperations.py
from typing import List
# 3397
class Solution:
    def maxDistinctElements(self, nums: List[int], k: int) -> int:
        
        nums.sort()
        before = float('-inf')
        distinct = 0

        for num in nums:
            lower_bound = num - k
            upper_bound = num + k

            if before < lower_bound:
                before = lower_bound
            else:
                before += 1
            
            if before <= upper_bound:
                distinct += 1
            else:
                before -= 1
        return distinct